File: 6kyu/JavaScript/mexicanWave.js
/*
Task
In this simple Kata your task is to create a function that turns a string into a Mexican Wave. You will be passed a string and you must return that string in an array where an uppercase letter is a person standing up. 
Rules
 1.  The input string will always be lower case but maybe empty.
 2.  If the character in the string is whitespace then pass over it as if it was an empty seat

Example
wave("hello") => ["Hello", "hEllo", "heLlo", "helLo", "hellO"]
*/

// METHOD 1
const wave = s => {
    let waveArray = []; // place to store the array
    
    // loop through string
    for(let i = 0; i < s.length; i++) {
    
        // if the current character is an empty space, continue onto next iteration
        if (s[i] == ' ') {
            continue;
        }
        // capitalize letter at index, then add the rest of the string to it
        let capitalizedSubstring = `${s[i].toUpperCase()}${s.substring(i+1)}`
    
        // if that string is equal in length the the original string
        if(capitalizedSubstring.length === s.length) {
            // add it to the waveArray array
            waveArray.push(capitalizedSubstring)

        // if that string is shorter than the original
        } else if (capitalizedSubstring.length < s.length) {
            // concatenate the beginning of the string before the capitalized letter
            let fullString = `${s.substring(0, i)}${capitalizedSubstring}`;
            waveArray.push(fullString);
        }
    }
    return waveArray;
  }
  
// METHOD 2
const wave = s => [...s]
  // map each char in string
  .map((e, i) => {
    // if the current element is not an empty space
    if (e !== ' ') {
     // take a slice of the string until the current index, then make the current index uppercase, and concatenate the rest of it with another slice
    return `${s.slice(0, i)}${s[i].toUpperCase()}${s.slice(i+1)
  }`}
    })
  // filter to make sure none of the elements are undefined (which they may be because of the if statement in our map)
  .filter(x => x !== undefined);

File: 6kyu/Python/mexicanWave.py
'''
Task
In this simple Kata your task is to create a function that turns a string into a Mexican Wave. You will be passed a string and you must return that string in an array where an uppercase letter is a person standing up. 
Rules
 1.  The input string will always be lower case but maybe empty.
 2.  If the character in the string is whitespace then pass over it as if it was an empty seat
Example
wave("hello") => ["Hello", "hEllo", "heLlo", "helLo", "hellO"]
'''

# METHOD 1
def wave(s):
    waveArr = [];
    
    i = 0; # index to count in for loop
    # for each character in the string
    for char in s:
        word = f'{s[0:i]}{char.upper()}{s[i+1:]}'
        waveArr.append(word)
        i += 1;
    
    # filter list to only items that are not equal to s (which is all lowercase) in order to skip the spaces
    filtered = list(filter(lambda e: e != s, waveArr))
    return filtered

# METHOD 2
def wave(s):
    # list comprehension to put it all on one line
    # concatenate the string up until i with the string from i on capitalized (so first letter of this substring is capitalized)
    # only do it if the current element is not an empty space
    return [f'{s[:i]}{s[i:].capitalize()}' for i in range(len(s)) if s[i] != ' ']

# METHOD 3: same as method 2 but with lambda (all on one line)
wave = lambda s: [f'{s[:i]}{s[i:].capitalize()}' for i in range(len(s)) if s[i] != ' ']
